File: lw7/slides/shapes/Rectangle.cpp
#include "Rectangle.h"
#include <cmath>

CRectangle::CRectangle(const PointD& leftTop, const PointD& rightBottom)
	: CSimpleShape()
	, m_leftTop(leftTop)
	, m_rightBottom(rightBottom)
{
}

RectD CRectangle::GetFrame()
{
	double width = std::abs(m_rightBottom.x - m_leftTop.x);
	double height = std::abs(m_rightBottom.y - m_leftTop.y);

	return RectD{ m_leftTop.x, m_leftTop.y, width, height };
}

void CRectangle::SetFrame(const RectD& rect)
{
	m_leftTop = PointD{ rect.left, rect.top };
	m_rightBottom = PointD{ rect.left + rect.width, rect.top + rect.height };
}

void CRectangle::DrawBehaviour(ICanvas& canvas) const
{
	PointD rightTop{ m_rightBottom.x, m_leftTop.y };
	PointD leftBottom{ m_leftTop.x, m_rightBottom.y };

	if (*GetOutlineStyle()->IsEnabled())
	{
		canvas.DrawLine(m_leftTop, rightTop);
		canvas.DrawLine(rightTop, m_rightBottom);
		canvas.DrawLine(m_rightBottom, leftBottom);
		canvas.DrawLine(leftBottom, m_leftTop);
	}

	if (*GetFillStyle()->IsEnabled())
	{
		canvas.FillPolygon(std::vector<PointD>{ m_leftTop, rightTop, m_rightBottom, leftBottom });
	}
}

File: lw9/src/model/shapes/Rectangle.js
import {Shape} from "./Shape"
import {Point} from "../types/Point.js"
import {Rect} from "../types/Rect.js"

class Rectangle extends Shape {
    /**
     * @param {Rect} rect
     */
    constructor(rect) {
        super()

        /**
         * @type {Point}
         * @private
         */
        this._leftTop = { x: rect.left, y: rect.top }

        /**
         * @type {Point}
         * @private
         */
        this._rightBottom = { x: rect.left + rect.width, y: rect.top + rect.height }
    }

    /**
     * @override
     * @private
     */
    _getTypeImpl() {
        return 'rectangle'
    }

    /**
     * @override
     * @private
     */
    _getRectImpl() {
        const left = this._leftTop.x
        const top = this._leftTop.y

        const width = Math.abs(this._rightBottom.x - this._leftTop.x)
        const height = Math.abs(this._rightBottom.y - this._leftTop.y)

        return {
            left,
            top,
            width,
            height,
        }
    }

    /**
     * @override
     * @private
     */
    _setRectImpl(rect) {
        this._leftTop = { x: rect.left, y: rect.top}
        this._rightBottom = { x: rect.left + rect.width, y: rect.top + rect.height }
    }
}

export {
    Rectangle,
}